File: 238.product-of-array-except-self.java
/*
 * @lc app=leetcode id=238 lang=java
 *
 * [238] Product of Array Except Self
 */

// @lc code=start
class Solution {
    public int[] productExceptSelf(int[] nums) {
        // The product of any prefix or suffix of nums
        // is guaranteed to fit in a 32-bit integer.
        int[] res = new int[nums.length];

        // get the product of i left, prodLeft
        for (int i = 0; i < nums.length; i++) {
            if (i == 0) {
                res[i] = 1;
            } else {
                res[i] = nums[i - 1] * res[i - 1];
            }
        }

        // get the oriduct of i right, prodRight
        // prodRight * prodLeft =
        // the product of all the elements of nums except nums[i]
        int prodRight = 1;
        for (int i = nums.length - 1; i >= 0; i--) {
            if (i == nums.length - 1) {
                res[i] *= prodRight;
            } else {
                prodRight *= nums[i + 1];
                res[i] *= prodRight;
            }
        }
        // Can you solve the problem in O(1) extra space complexity?
        // (The output array does not count as extra space
        // for space complexity analysis.)

        return res;
    }
}
// @lc code=end

File: 238.product-of-array-except-self.py
#
# @lc app=leetcode id=238 lang=python3
#
# [238] Product of Array Except Self
#


# @lc code=start
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # The input is generated such that answer[i] is guaranteed to fit in a 32-bit integer.
        # O(n), O(1)
        n = len(nums)
        answers = [1] * n
        for i in range(1, n):
            answers[i] = answers[i - 1] * nums[i - 1]

        # x=product of nums[i:]
        x = 1
        for i in range(n - 2, -1, -1):
            x *= nums[i + 1]
            answers[i] *= x

        return answers


# @lc code=end
